### DISCUSSION TITLE
Launching from Ubuntu pulsar.desktop fails

### DISCUSSION THREAD

**Discussion by croaker000** [pulsar-edit/.github#161] [Pulsar Support] [answered]
Starting Pulsar from the Ubuntu dashboard fails to start the program. Nothing happens. Launching from a terminal window by typing "pulsar" works fine.

The dashboard method reports this in the log. Why is Pulsar trying to start Bluetooth (Bluez)? I don't have bluetooth on this box.

```
Apr  4 15:46:12 bomanz dbus-daemon[2603]: [system] Activating via systemd: service name='org.bluez' unit='dbus-org.bluez.service' requested by ':1.1479' (uid=1000 pid=1038638 comm="/opt/Pulsar/pulsar" label="unconfined")
Apr  4 15:46:12 bomanz systemd[1]: Bluetooth service was skipped because of a failed condition check (ConditionPathIsDirectory=/sys/class/bluetooth).
Apr  4 15:46:13 bomanz systemd[1]: Starting Process error reports when automatic reporting is enabled...
Apr  4 15:46:13 bomanz systemd[1]: Started crash report submission.
Apr  4 15:46:13 bomanz whoopsie[1038770]: [15:46:13] Using lock path: /var/lock/whoopsie/lock
Apr  4 15:46:13 bomanz systemd[1]: whoopsie.service: Deactivated successfully.
Apr  4 15:46:13 bomanz whoopsie-upload-all[1038769]: INFO:root:Collecting info for /var/crash/_opt_Pulsar_pulsar.1000.crash...
Apr  4 15:46:13 bomanz systemd[1]: Started crash report submission.
Apr  4 15:46:13 bomanz whoopsie[1038781]: [15:46:13] Using lock path: /var/lock/whoopsie/lock
Apr  4 15:46:13 bomanz systemd[1]: whoopsie.service: Deactivated successfully.
Apr  4 15:46:15 bomanz pulsar.desktop[1038638]: [1038638:0404/154615.092825:FATAL:gpu_data_manager_impl_private.cc(415)] GPU process isn't usable. Goodbye.
Apr  4 15:46:15 bomanz kernel: [439706.601472] traps: Chrome_IOThread[1038701] trap int3 ip:55708c3aacd8 sp:7f7f9d0763e0 error:0 in pulsar[55708c17f000+6271000]
Apr  4 15:46:17 bomanz whoopsie-upload-all[1038769]: WARNING: hook failed for processing /var/crash/_opt_Pulsar_pulsar.1000.crash:
Apr  4 15:46:17 bomanz whoopsie-upload-all[1038769]: ERROR: processing /var/crash/_opt_Pulsar_pulsar.1000.crash: Compressed file ended before the end-of-stream marker was reached
Apr  4 15:46:17 bomanz systemd[1]: Started crash report submission.
Apr  4 15:46:17 bomanz systemd[1]: apport-autoreport.service: Deactivated successfully.
Apr  4 15:46:17 bomanz systemd[1]: Finished Process error reports when automatic reporting is enabled.
Apr  4 15:46:17 bomanz systemd[1]: apport-autoreport.service: Consumed 4.752s CPU time.
Apr  4 15:46:17 bomanz whoopsie[1039582]: [15:46:17] Using lock path: /var/lock/whoopsie/lock
Apr  4 15:46:17 bomanz systemd[1]: whoopsie.service: Deactivated successfully.
```

**Comment by Daeraxa** [comment]
What package are you using? Was it an automatic .desktop file (e.g. from the .deb) or was it made yourself or by a tool like appimagelauncher if using one of the others.

**Reply by croaker000** [reply]
It was the deb package download from the Pulsar site. The .desktop was the one included in the deb.

I can't see anything untoward in it.

**Comment by confused-Techie** [comment]
I'd have to investigate but is it possible that it's Electrons Chromium that tries to access bluetooth? The same way that the built in Chromium of Electron will fail on Linux systems without CUPS.

Are you able to run any other Electron apps? Think Discord, or Bitwardens Desktop Client

**Reply by croaker000** [reply]
That was my initial thought too but don't know enough about how electron works to be sure.
I'm a migrating Atom user so yes, other Electron things work.

**Reply by confused-Techie** [reply]
If you don't have any packages installed, there's nothing within Pulsar itself that would be trying to use bluetooth.

So I'm honestly still wanting to blame Chromium for doing this. Other than Atom have you run any other Electron Apps? Since Atom used a very very old version, and we are to, but it's newer than what Atom had.

It's possible this change is because of the upgrades? I'll see if I can find anything to confirm that.

**Reply by confused-Techie** [reply]
So I have found here that someone did report this previously on the Atom forms (back when those existed) and was linked to by another Electron App where someone is reporting the same error on [SignalApp](https://github.com/signalapp/Signal-Desktop/issues/3975)

**Reply by confused-Techie** [reply]
Also [here's](https://www.reddit.com/r/chrome/comments/8k8t1t/why_is_chrome_trying_to_access_bluetooth/) the same error reported for Google Chrome. So I suppose an actual good test would be to see if Google Chrome causes the same issue.

I'm going to see if there's any flags that exist to disable Chromium trying to access Bluetooth, but went looking for the same thing when it came to CUPS and there wasn't any so I'm not super hopeful.

It's possible the best solution here is to enable bluetooth like seen on this [report](https://bbs.archlinux.org/viewtopic.php?id=155714) of the same error, and see if it lets you launch the application.

**Reply by confused-Techie** [reply]
Lastly it seems it might be possible to start Pulsar with the following flags `--disable-features=enable-web-bluetooth`.

A warning I don't have a Linux machine to test this on, but maybe hopefully it can work?

**Reply by croaker000** [reply]
I have no bluetooth controllers but switching the service back on:

```
mart@bomanz:~$ systemctl status dbus-org.bluez.service
○ bluetooth.service - Bluetooth service
     Loaded: loaded (/lib/systemd/system/bluetooth.service; enabled; preset: enabled)
     Active: inactive (dead)
       Docs: man:bluetoothd(8)
```

Still fails:

```
Apr  5 01:03:45 bomanz systemd[12760]: Started Application launched by gnome-shell.
Apr  5 01:03:45 bomanz dbus-daemon[2603]: [system] Activating via systemd: service name='org.bluez' unit='dbus-org.bluez.service' requested by ':1.1615' (uid=1000 pid=1458788 comm="/opt/Pulsar/pulsar" label="unconfined")
Apr  5 01:03:45 bomanz systemd[1]: Bluetooth service was skipped because of a failed condition check (ConditionPathIsDirectory=/sys/class/bluetooth).
```
Presumably because I don't have any bluetooth controllers or devices?

With the service enabled or disabled, `--disable-features=enable-web-bluetooth` doesn't make any difference

NB: I found this, it looks like VSCode has gone through a similar trauma.
https://github.com/microsoft/vscode/issues/134461

**Reply by confused-Techie** [reply]
Yeah looks like that issue is still occurring for VSCode as well. Which they link the Electron code that causes it.

So in that case might not be much we can do from the Pulsar side, as the issue is upstream. But I'm surprised enabling it doesn't cause the issue to go away. As in starting up the bluetooth service

**Reply by croaker000** [reply]
Maybe because the service is still "inactive" despite being "enabled"? I have no bluetooth controller so the service can't hook into anything to make it "active"...

Odd that Atom works fine though - maybe that's why they used a very old version of chromium?

**Reply by confused-Techie** [reply]
So considering this potentially means Bluetooth is required to run any Electron App (which is just crazy it hasn't been fixed yet).

I've found that someone did open an issue for this on Electron [here](https://github.com/electron/electron/issues/36831). Which was closed because Electron was doing this to match what Chromium does. Although I can't find any issues created for [Chromium](https://bugs.chromium.org/p/chromium/issues/list?q=bluez&can=1) of anyone mentioning this being an issue. So either I'm not able to find it, or nobody has brought it up with them?

**Reply by confused-Techie** [reply]
It's possible this contributed to them using an older Electron (thus Chromium), but really the upgrade path is a very difficult one. But required if we want to have modern security, features, and performance. I think really the reason they never did it is they wouldn't give anyone the time to do it. Atom was neglected for about three years, and 15 odd Electron versions. So things are very far behind.

**Reply by confused-Techie** [reply]
Really I think the best thing we could do would be to see if Chromium works. That way we can at least try to determine if this issue has been fixed in later chromium builds.

**Reply by croaker000** [reply]
I'm on Ubuntu 22.10 which only has direct access to the snap version <yuck!>

Using: https://fosspost.org/chromium-deb-package-ubuntu-22-04 , I've installed the deb version which runs from the .desktop file and reports:

Version 108.0.5359.40 (Developer Build) Ubuntu 22.10 (64-bit)

Hope that helps...

**Reply by croaker000** [reply]
One thing I still haven't figured out is why the .desktop file route fails from the dashboard but typing "pulsar" from a terminal window or Ubuntu's ALT-F2 "Run a command" window works fine.

**Reply by Spiker985** [reply, marked as answer]
One thing to check would be if adding `--no-sandbox` to the .desktop helps for some god forsaken reason

It's something that's added inside the pulsar.sh, that is missing from the .desktop file that electron-builder creates

**Reply by croaker000** [reply]
That did it! Well done. Sorted. Thanks.

**Reply by Spiker985** [reply]
Son of a- Alright, I'll see if I can PR that into the electron builder config

**Reply by confused-Techie** [reply]
> Son of a- Alright, I'll see if I can PR that into the electron builder config

It might be worth mentioned that when investigate the other day it looked like it was possible to add some flags during runtime, right before the application window is created, at this point if we need it on all Linux systems, then maybe it's possible to add it there instead of trying to shove it into every possible way Pulsar could be launched?

Also @croaker000 had no clue it worked via the CLI, so glad a solution was found. But really interesting that `--no-sandbox` disables the requiring of core components like that

**Reply by Spiker985** [reply]
> > Son of a- Alright, I'll see if I can PR that into the electron builder config
> 
> It might be worth mentioned that when investigate the other day it looked like it was possible to add some flags during runtime, right before the application window is created, at this point if we need it on all Linux systems, then maybe it's possible to add it there instead of trying to shove it into every possible way Pulsar could be launched?
> 
> Also @croaker000 had no clue it worked via the CLI, so glad a solution was found. But really interesting that `--no-sandbox` disables the requiring of core components like that

If we could inject it, that would also presumably fix the instances where the app image doesn't function properly as well

Edit: The problem is, that flag is for Electron and I'm unsure if we can set it that late in the startup

**Reply by confused-Techie** [reply]
@Spiker985 yeah looks like we could do:

```javascript
app.commandLine.appendSwitch("no-sandbox");
```

And as long as it happens before the `ready` event we should be good.

[Source](https://www.electronjs.org/docs/latest/api/command-line-switches)
